### DISCUSSION TITLE
Qualification Workflow PRs/Issues/Branches

### DISCUSSION THREAD

**Discussion by kodesiba** [Gilead-Public/gsm#576] [Ideas]
# Current Process
Develop Features and Bug Fixes on fix branches down stream from the dev branch.  Developer then puts in a PR to dev after updating docs and non-qualification unit tests.  This PR will still not have passing actions because of the out of date qualification tests living in the tests/testthat folder.  The Qualifier then checks out a branch downstream from the developer's fix branch and updates qualification tests then puts in a PR into the fix branch.  This is the sticking point to me as it impedes development and leaves PRs to dev unmerged while the qualifier works on the qualification.  

## Pros

- Branches are consolidated into one update
- Qualification tests contribute to code coverage (possibly good thing, possibly sketchy depending on who is looking)

## Cons
 - Branches can get left in qualification limbo while they wait for the tests to be updated
 - PRs to dev might be merged without fully passing checks to speed up development
 - Development being merged into dev is likely to be also contingent on qualification test updates to meet best practices

# Proposed Updates
Move qualification tests into a tests/qualification folder.  This will make it so developers only need to have unit tests and docs updated before they put through a pull request to dev.  Qualification can be done in larger chunks as larger pieces of development is done or if multiple updates will be put through that impact these tests.  Developer would then put in an issue to outline the updates made that would need to be qualified for the qualifier to make a branch based on.

## Pros

- Development and qualification become separate workflows both done downstream from dev and merged into dev
- Qualification no longer is a blocker to development PRs getting merged into dev with clean checks

## Cons

 - Extra branches and issues made 
 - Qualification tests don't contribute to code coverage (possibly good thing, possibly sketchy depending on who is looking)

I'm fine with both but I think that the outlined updates would improve the qualification process and make it more streamlined and asynchronous.  Let me know if anyone has thoughts on this, if we want to move forward with this it should be a pretty easy update from the qualification end.

@jwildfire @mattroumaya @samussiah @ge-li @gwu05

**Comment by samussiah** [comment]
Makes sense to me!  Is it possible to set up a workflow specifically for qualification fix branches into dev that requires both unit and qualification tests to pass?

**Comment by jwildfire** [comment]
> Qualification tests don't contribute to code coverage (possibly good thing, possibly sketchy depending on who is looking)

I do kind of like having everything together in `/test/testthat`. In addition to the comment above - I think the biggest advantage of the current system is that there is a very tight feedback loop between dev & QC. I like that the QC tests currently run as part of the standard testing workflow (`devtools::test()/check()` which it tells devs very quickly if they break something in qualification. 

That said, I'm ok updating the process to have the qualification tests run separately as part of the release process if that's the groups preference.

**Reply by ge-li** [reply]
Would it be possible to create a workflow similar to `devtools::test()/check()` under the proposed QC updates? @kodesiba

**Reply by kodesiba** [reply]
We can keep the check on the PR to dev but pull it out as it's own action since it can be in it's own folder.  For developers the workflow could be something like PR to dev, CMD check passing is a must, if qualification check fails make an issue to update QC tests when you merge it to dev.
